### PR TITLE
Add a pre-configured guard & default server middleware.

### PR DESCRIPTION
**Pull request by DFurnes**
#### Changes
This pull request adds pre-configured server support for Northstar-issued JWT tokens:

🛡 Adds a `GatewayGuard`, which allows Laravel's built-in authentication to work for API requests made with Northstar tokens (via built-in middleware or `auth()->id()`, etc.)

👤 Adds a `GatewayUserProvider`, which allows resource servers to request user information from tokens (so `auth()->user()` would return the authenticated user's profile information).

🍔 Adds `RequireUser`, `RequireRole`, and `RequireScope` middleware which are pre-configured to protect routes using JWT tokens.

🎫 Adds `token()` helper which can grab details from the JWT on the current request on-demand. For example, this might be used to attach client details as a "source" with `token()->client`.

#### How should this be reviewed?
I'm going to add a few little test cases, and there's also a [corresponding Rogue pull request](https://github.com/DoSomething/rogue/compare/gatekeeper) demonstrating integration into a real Laravel application!

### CODE REVIEW

File: src/Laravel/helpers.php [hunk 1]
```diff
@@ -18,4 +19,14 @@ function gateway($client)
 
         return app(Northstar::class);
     }
+
+    /**
+     * Return the current OAuth token.
+     *
+     * @return Token
+     */
+    function token()
```
Comment (weerd) [inline src/Laravel/helpers.php:28]: useful!

File: src/Server/GatewayUserProvider.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    /**
+     * Retrieve a user by their unique identifier and "remember me" token.
+     *
+     * @param  mixed   $identifier
+     * @param  string  $token
+     * @return \Illuminate\Contracts\Auth\Authenticatable|null
+     */
+    public function retrieveByToken($identifier, $token)
+    {
+        throw new InvalidArgumentException('Remember tokens not supported via Gateway.');
```
Comment (weerd) [inline src/Server/GatewayUserProvider.php:34]: womp womp!

File: tests/Server/Middleware/RequireRoleTest.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
+<?php
+
+use Carbon\Carbon;
+use DoSomething\Gateway\Server\Token;
+use DoSomething\Gateway\Server\Middleware\RequireRole;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+
+class RequireRoleTest extends TestCase
+{
+    /** @test */
```
Comment (weerd) [inline tests/Server/Middleware/RequireRoleTest.php:10]: I don't think you need the `/** @test */` comment if the method begins with `test...()`. That being said didn't we move to not naming tests w/ prefix `test` and want to move to `it_can_` or whatever? Also not sure if makes sense to change if rest of package already does `test` prefix. Just random thoughts 🤔

**Review by weerd — APPROVED** [review]
I suck at OAuth but the PHP looks snazzy! 🌮

File: tests/Server/Middleware/RequireRoleTest.php (hunk 3 above)
Comment (DFurnes) [reply tests/Server/Middleware/RequireRoleTest.php:10]: Yeah I think you're right! I was considering writing these as `it_should_...` style, but decided it made a bit more sense to hold off & maybe do them all at once in another PR.

**Pull request merged**